pkg/evaluation/evaluator.go

```diff
@@ -0,0 +1,355 @@
+package evaluation
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/harness/ff-golang-server-sdk/log"
+	"github.com/harness/ff-golang-server-sdk/rest"
+)
+
+const (
+	oneHundred = 100
+
+	segmentMatchOperator   = "segmentMatch"
+	matchOperator          = "match"
+	inOperator             = "in"
+	equalOperator          = "equal"
+	gtOperator             = "gt"
+	startsWithOperator     = "starts_with"
+	endsWithOperator       = "ends_with"
+	containsOperator       = "contains"
+	equalSensitiveOperator = "equal_sensitive"
+)
+
+// Query provides methods for segment and flag retrieval
+type Query interface {
+	GetSegment(identifier string) (rest.Segment, error)
+	GetFlag(identifier string) (rest.FeatureConfig, error)
+}
+
+// Evaluator engine evaluates flag from provided query
+type Evaluator struct {
+	query Query
+}
+
+// NewEvaluator constructs evaluator with query instance
+func NewEvaluator(query Query) *Evaluator {
+	return &Evaluator{
+		query: query,
+	}
+}
+
+func (e Evaluator) evaluateClause(clause *rest.Clause, target *rest.Target) bool {
+	if clause == nil {
+		return false
+	}
+
+	operator := clause.Op
+	if operator == "" {
+		return false
+	}
+
+	attrValue := getAttrValue(target, clause.Attribute)
+	if attrValue.IsNil() {
+		return false
+	}
+
+	object := attrValue.String()
+	values := clause.Values
+	if len(values) == 0 {
+		return false
+	}
+
+	value := values[0]
+
+	switch operator {
+	case startsWithOperator:
+		return strings.HasPrefix(object, value)
+	case endsWithOperator:
+		return strings.HasSuffix(object, value)
+	case matchOperator:
+		found, err := regexp.MatchString(value, object)
+		if err != nil || !found {
+			return false
+		}
+		return true
+	case containsOperator:
+		return strings.Contains(object, value)
+	case equalOperator:
+		return strings.EqualFold(object, value)
+	case equalSensitiveOperator:
+		return object == value
+	case inOperator:
+		for _, val := range values {
+			if val == object {
+				return true
+			}
+		}
+		return false
+	case gtOperator:
+		return object > value
+	case segmentMatchOperator:
+		return e.isTargetIncludedOrExcludedInSegment(&values, target)
```

Comment:
Not sure this needs to be passed by ptr.  The slide header gets passed as a copy, but it will point to the same underlying array.

Reply:
I just followed rest types, Yes I know about slicing and underlaying arrays, I will change method signatures